samples/99_zil_interpreter/app/main.rb

```diff
@@ -7,39 +7,107 @@
 
 def tick(args)
   setup(args) if args.tick_count.zero?
-  # TODO:
-  # - Render history
-  # - Collect input
-  # - Call send_input with input when pressing Enter
+  # Render history
+  render_state(args)
+  
+  # Collect input
+  handle_input(args)
+  
+  # Call send_input when pressing Enter
+  send_input(args, args.state.input)
+
+  
   $gtk.request_quit unless $interpreter.alive?
 end
 
 def setup(args)
+  args.state.text_color = [240, 255, 255]
+  args.state.bg_color = [0, 0, 0]
+  args.state.input = ''
+  args.state.text_history = []
+  args.state.current_line = 0
   context = build_zil_context(args)
   args.state.zil_context = context
   $interpreter = Fiber.new {
     context.globals[:GO].call [], context
   }
   $interpreter.resume # Initial processing until first Fiber.yield
-  process_outputs args, context.outputs # Process welcome message if existing
+  process_outputs(args, context.outputs) # Process welcome message if existing
 
   # TODO:
   # Add other setup if necessary
 end
 
-# Call this method with the input after pressing enter
+def handle_input(args)
+  if args.inputs.keyboard.key_down.backspace
+    args.state.input.chop!
+  else
+    if args.inputs.text[0]
+      args.state.input << args.inputs.text[0]
+      args.state.current_line = 0
+    end
+  end
+
+  # Scrolling
+  if args.inputs.mouse.wheel
+    args.state.current_line += args.inputs.mouse.wheel.y
+
+    # Stop the player from scrolling past the beginning or end
+    max_possible_line = [0, args.state.text_history.length - 33].max
+    args.state.current_line = args.state.current_line.clamp(0, max_possible_line)
+
+  # Jump to the present
+  elsif args.inputs.keyboard.key_down.escape
+    args.state.current_line = 0
+  end
+end
+
+# Called with the input after pressing enter
 def send_input(args, input)
-  $interpreter.resume input
+  if args.inputs.keyboard.key_down.enter && args.state.input != ''
```

Comment:
I think that check makes more sense in the main tick method - 

maybe like this?
```rb
def tick(args)
  ...
  send_input(args, args.state.input) if args.inputs.keyboard.key_down.enter
  ...
end

def send_input(args, input)
  return if input.empty?
  
  ....
end
```